lib/_http_outgoing.js

```diff
@@ -123,6 +129,8 @@ function OutgoingMessage() {
   this._removedContLen = false;
   this._removedTE = false;
 
+  this[kBytesWritten] = 0;
+  this[kContentLength] = null;
```

Comment:
This seems to duplicate `this._contentLength`, can't we reuse it instead?

Reply:
Actually, that's exactly what I wanted to do. Still, I was a bit unsure if updating `this._contentLength`'s value would cause some other functionality to break cause Ronag said to symbols in some previous PR for this thing.

I will make this change and double-check if it works the same way as expected, will be much more cleaner and efficient.

Reply:
I don't think there should be any problem reusing `this._contentLength` since we're not changing the behavior of Content-Length tracking at all, it should be keeping track of the bytes written and comparing that to the Content-Length if it was set.

Reply:
@mscdex Yes that makes sense, I've removed the previous way of checking for the content-length in the favour of doing this in the `matchHeader` function,

```js
  case 'content-length':
      state.contLen = true;
      self._contentLength = value;
```

There was some issue initially because the `_contentLength` was just coming out as null so setting it like so fixed that.

Reply:
Also, something I am wondering about, I understood that all the methods at last delegate the writing work to the `_writeRaw` method so sometimes it can also be called with the headers instead of the body payload in which case this calculation doesn't seem right cause AFAIK `content-length` is for body right?

```js
 const buffLength = Buffer.byteLength(data, encoding);
      if (buffLength + this[kBytesWritten] > this._contentLength) {
          throw new ERR_HTTP_CONTENT_LENGTH_MISMATCH(buffLength + this[kBytesWritten], this._contentLength);
      }
```

I think this calculation should take into account that the `total payload size = chunk's byte length - header's byte length`

 So according to you what might be the efficient way to check if the chunk only contains headers in which case we can skip it

Reply:
I fixed this by doing
```js
  if (conn && !conn.destroyed && !this._removedContLen && typeof this._contentLength === "number") {
    const headerLength = conn._httpMessage._hasBody ? conn._httpMessage._header.length : 0;
    const buffLength = Buffer.byteLength(data, encoding) % headerLength;
    if (buffLength + this[kBytesWritten] > this._contentLength) {
      throw new ERR_HTTP_CONTENT_LENGTH_MISMATCH(buffLength + this[kBytesWritten], this._contentLength);
    }
    this[kBytesWritten] += buffLength;
  }
```

@mscdex Why does the `OutgoingMessage.end` run twice on a single request? and the `._contentLength`is null the first time

```
CONTENT LENGTH IN "res.end"
 _contentLength = null
 getHeader('content-length') = undefined
 chunk = undefined

******************************************************

CONTENT LENGTH IN "res.end"
 _contentLength = 5
 getHeader('content-length') = 5
 chunk = undefined

******************************************************
```

 I think I need to understand the flow of the request/response cycle better, any specific parts in the docs that I should read?

Reply:
Now this makes sense https://nodejs.org/dist/latest-v18.x/docs/api/http.html#class-httpoutgoingmessage